src/Interpreters/QueryMetricLog.cpp

```diff
@@ -0,0 +1,200 @@
+
+    auto context = getContext();
+    const auto & process_list = context->getProcessList();
+    status.task = context->getSchedulePool().createTask("QueryMetricLog", [this, &process_list, query_id] {
+        auto current_time = std::chrono::system_clock::now();
+        const auto query_info = process_list.getQueryInfo(query_id, false, true, false);
+        if (!query_info)
+        {
+            LOG_TRACE(logger, "Query {} is not running anymore, so we couldn't get its QueryInfo", query_id);
+            return;
+        }
+
+        auto elem = createLogMetricElement(query_id, *query_info, current_time);
+        if (elem)
+            add(std::move(elem.value()));
+        else
+            LOG_TRACE(logger, "Query {} finished already while this collecting task was running", query_id);
+    });
+
+    status.task->scheduleAfter(interval_milliseconds);
+
+    std::lock_guard lock(queries_mutex);
+    queries.emplace(query_id, std::move(status));
+}
+
+void QueryMetricLog::finishQuery(const String & query_id, QueryStatusInfoPtr query_info)
+{
+    std::unique_lock lock(queries_mutex);
+    auto it = queries.find(query_id);
+
+    /// finishQuery may be called from logExceptionBeforeStart when the query has not even started
+    /// yet, so its corresponding startQuery is never called.
+    if (it == queries.end())
+        return;
+
+    if (query_info)
+    {
+        auto elem = createLogMetricElement(query_id, *query_info, std::chrono::system_clock::now(), false);
+        if (elem)
+            add(std::move(elem.value()));
+    }
+
+    /// The task has an `exec_mutex` locked while being executed. This same mutex is locked when
+    /// deactivating the task, which happens automatically on its destructor. Thus, we cannot
+    /// deactivate/destroy the task while it's running. Now, the task locks `queries_mutex` to
+    /// prevent concurrent edition of the queries. In short, the mutex order is: exec_mutex ->
+    /// queries_mutex. Thus, to prevent a deadblock we need to make sure that we always lock them in
+    /// that order.
+    {
+        /// Take ownership of the task so that we can destroy it in this scope after unlocking `queries_lock`.
+        auto task = std::move(it->second.task);
+
+        /// Build an empty task for the old task to make sure it does not lock any mutex on its destruction.
+        it->second.task = {};
+
+        /// Ensure `queries_mutex` is unlocked before calling task's destructor at the end of this
+        /// scope which will lock `exec_mutex`.
+        lock.unlock();
+    }
+
+    lock.lock();
+    queries.erase(query_id);
+}
+
+std::optional<QueryMetricLogElement> QueryMetricLog::createLogMetricElement(const String & query_id, const QueryStatusInfo & query_info, TimePoint current_time, bool schedule_next)
+{
+    LOG_DEBUG(logger, "Collecting query_metric_log for query {}. Schedule next: {}", query_id, schedule_next);
```

Comment:
Debug logging looks annoying, maybe `LOG_TEST` will be enough?

Reply:
I'll demote it to `LOG_TRACE` in another PR where I'm working on some improvements for this :+1: 
I'm choosing `trace` instead of `test` because that's the default level for the CI and it's helped me a lot to debug issues only seen on the CI

Reply:
After giving it another thought `LOG_TRACE` still can be too much, especially due to it is the default and what in general users use, maybe CI should try to use `level=test` one more time

I remember that previously there was some issues and it has been disabled here https://github.com/ClickHouse/ClickHouse/pull/48424#issuecomment-1637722113 (cc @kssenii )

Reply:
I also gave it some thought and decided to add a new setting to enable this sort of traces just for the queries that are of our interest: [`955f537` (#71473)](https://github.com/ClickHouse/ClickHouse/pull/71473/commits/955f537bd5ef2f4a29717ac4999ce2af47b4c039)

So, for the stateless test it's enabled through `query_metric_log_debug` to help debugging it, but it doesn't print any noise during normal behavior. Does it work for you?

Reply:
Yep, thank you!

Reply:
> Though settings just for this looks too much

Yep, completely agree. After giving it another thought, I decided to keep the logs for a while to ensure everything is good on the CI, but remove them before the release of the next version.